### PR TITLE
🌱 Ironic-deployment: Use kustomize components

### PR DESCRIPTION
**Pull request by lentzi90**
**What this PR does / why we need it**:

This changes the ironic-deployment kustomizations to make use of kustomize components. The goal is to simplify the structure and make it easier to add new configurations in the future (e.g. mariadb/sqlite), as well as make the kustomizations more useful to downstream users (i.e. they should be useful on their own or as kustomize base for building downstream overlays).

The main advantage of components over overlays for this use case is that components can be mixed and matched, whereas overlays cannot. For example, you cannot "add" together the TLS and basic-auth overlays since there will be conflicts, which is the reason for the current complex structure.

With components we get this:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: baremetal-operator-system
namePrefix: baremetal-operator-

resources:
- ../../base
- ../../../config/namespace

components:
- ../../components/basic-auth
- ../../components/tls
- ../../components/keepalived
```

With overlays we instead end up with a chain of overlays on top of each other:

```yaml
# basic-auth/tls/keepalived
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: baremetal-operator-system
resources:
- ../../../tls/keepalived
...
---
# tls/keepalived
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: baremetal-operator-system
resources:
- ../../keepalived
- ../../certmanager
...
---
# keepalived
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: baremetal-operator-system
namePrefix: baremetal-operator-
resources:
- ../../config/namespace
- ../ironic
...
```

### CODE REVIEW

**Comment by dtantsur** [comment]
I don't think I understand Kustomize enough to vote here, but I absolutely ❤️  the idea of this change.

**Comment by lentzi90** [comment]
Note: This is tested in https://github.com/metal3-io/metal3-dev-env/pull/1086
Since metal3-dev-env is using these kustomizations and the deploy script, we need that PR merged at the same time as this to keep the CI happy.

/hold

**Comment by lentzi90** [comment]
Updated docs to reflect the new structure.

**Review by Rozzii — APPROVED** [review]

File: docs/deploying.md
Comment (kashifest) [inline docs/deploying.md:135]: temp?

File: ironic-deployment/components/basic-auth/auth.yaml
Comment (kashifest) [inline ironic-deployment/components/basic-auth/auth.yaml:31]: Please check the order, should it be that we might need to keep this here and remove the other one?

File: docs/deploying.md
Comment (lentzi90) [reply docs/deploying.md:135]: Good find. I will add an explanation for this in the docs and probably remove it from here since it is a temporary folder (removed on `make clean`).

File: tools/deploy.sh
Comment (lentzi90) [reply tools/deploy.sh:115]: You mean deploying with/without keepalived? No it is still there but there is no explicit "default". We just add the keepalived components if needed, otherwise we leave it.
https://github.com/metal3-io/baremetal-operator/blob/b7251a0288d581995c0f7897c96dcc822166cd82/tools/deploy.sh#L120-L122

File: tools/deploy.sh
Comment (lentzi90) [reply tools/deploy.sh:125]: Good point! I'm reworking this a bit now, hopefully it will be less duplication of conditions. At first my idea was to keep ironic and BMO separate to make it clearer, but I think it didn't really work well.

File: tools/deploy.sh
Comment (furkatgofurov7) [reply tools/deploy.sh:115]: > You mean deploying with/without keepalived?

yeah, basically when `DEPLOY_KEEPALIVED == false`

File: ironic-deployment/components/basic-auth/auth.yaml
Comment (lentzi90) [reply ironic-deployment/components/basic-auth/auth.yaml:31]: I got confused myself, but this is actually correct. Added comments now to explain.
It goes like this: Inspector needs the credentials for accessing ironic, so it has the `ironic-auth-config`. This matches the `ironic-htpasswd` that Ironic has. And the other way: Ironic need the credentials for accessing inspector, so it has the `ironic-inspector-auth-config`. This matches the `ironic-inspector-htpasswd`.

File: tools/deploy.sh
Comment (lentzi90) [reply tools/deploy.sh:115]: Yes so if `DEPLOY_KEEPALIVED` is false (or anything other than true), then we skip adding the keepalived component. This should give us the exact same as before :slightly_smiling_face:

File: tools/deploy.sh
Comment (lentzi90) [reply tools/deploy.sh]: Done now, PTAL!

Comment (furkatgofurov7) [reply tools/deploy.sh]: Thanks, it looks good to me 👍🏼

**Review by furkatgofurov7 — COMMENTED** [review]
Assuming we have tested this patch properly with the dev env PR, this LGTM. 

/lgtm

**Comment by lentzi90** [comment]
> Assuming we have tested this patch properly with the dev env PR, this LGTM.

Thanks! Yes it has been tested with the latest changes in the other PR at least for the "normal" configurations that we use. I have not run tests for configurations we don't normally test (e.g. no TLS, no basic auth). But I have verified that the kustomizations are equivalent to those we already have for all combinations.

File: tools/deploy.sh
Comment (tuminoid) [inline tools/deploy.sh:121]: The specific combination of `[` and `==` should be avoided as `==` in non-Bash shell is undefined. So called bashism. For POSIX compliant scripting, use single `=` with single `[`. If you need to use Bash-only features, then make it `[[`. Bashisms may trigger unwanted errors in systems where default shell isn't Bash (for example Ubuntu often has dash).

Since this patch is not adding the line, feel free to not fix it in this patch, but avoid bashisms in future.

**Review by tuminoid — COMMENTED** [review]
Making a note of existing bashisms, feel free to ignore for this patch.

**Comment by kashifest** [comment]
@honza @zaneb @dtantsur @andfasano We are ready to take this patch in. If you dont have any adverse effect of this  in your downstream, I will approve this.

**Comment by dtantsur** [comment]
We don't use these files in OpenShift.

**Comment by kashifest** [comment]
> We don't use these files in OpenShift.

Thanks for the info.
/approve

**Comment by lentzi90** [comment]
Thanks for the reviews!
I think that we should be able to get green tests on this PR if we merge https://github.com/metal3-io/metal3-dev-env/pull/1086 first. However, other PRs will have failing tests then until this is merged.

**Pull request merged**
